src/pathTemplate.ts

```diff
@@ -66,45 +46,46 @@ export class PathTemplate {
    * @throws {TypeError} if path can't be matched to this template
    */
   match(path: string): Bindings {
-    const pathSegments = path.split('/');
+    let pathSegments = path.split('/');
+
     const bindings: Bindings = {};
-    let segmentCount = this.size;
-    let current: string;
-    let index = 0;
-    this.segments.forEach(segment => {
-      if (index > pathSegments.length) {
-        return;
+    if (pathSegments.length !== this.segments.length) {
+      // if the path contains a wildcard, then the length may differ by 1.
+      if (!this.data.includes('**')) {
+        throw new TypeError(
+          `This path ${path} does not match path template ${this.data}, the number of parameters is not same.`
+        );
+      } else if (pathSegments.length !== this.segments.length + 1) {
+        throw new TypeError(
+          `This path ${path} does not match path template ${this.data}, the number of parameters is not same with one wildcard.`
+        );
       }
-      if (segment.kind === extras.BINDING) {
-        current = segment.literal;
-      } else if (segment.kind === extras.TERMINAL) {
-        if (segment.literal === '*') {
-          bindings[current] = pathSegments[index];
-          index += 1;
-        } else if (segment.literal === '**') {
-          const size = pathSegments.length - segmentCount + 1;
-          segmentCount += size - 1;
-          bindings[current] = pathSegments.slice(index, index + size).join('/');
-          index += size;
-        } else if (segment.literal === pathSegments[index]) {
-          index += 1;
-        } else {
-          const msg = util.format(
-            "mismatched literal (index=%d): '%s' != '%s'",
-            index,
-            segment.literal,
-            pathSegments[index]
+    }
+    for (
+      let index = 0;
+      index < this.segments.length && pathSegments.length > 0;
+      index++
+    ) {
+      if (this.segments[index] !== pathSegments[0]) {
+        if (!this.segments[index].includes('*')) {
+          throw new TypeError(
+            `segment does not match, ${this.segments[index]} and  ${pathSegments[index]}.`
           );
-          throw new TypeError(msg);
+        } else if (this.segments[index].includes('*')) {
+          const segment = this.segments[index];
+          const variable = segment.match(/(?<={)[$0-9a-zA-Z_]+(?==.*})/g);
```

Comment:
Just in case, please use minimal quantifier here (`.*?` instead of `.*`). In an (invalid) case of a path template having two `=......}` occurrences, we want to make sure `.*` eats the minimum possible match, hence `.*?`.

Reply:
Thanks! applying the changes now.
This PR is only for rewriting the logic, and make sure all original unit tests pass. I will make a separate one to add non-slash-resource logic.